src/bin/consumer/errors.rs

```diff
@@ -24,14 +27,39 @@ impl AssignmentCallbacks for EmptyCallbacks {
 struct Next {
     destination: TopicOrPartition,
     producer: KafkaProducer,
+    last_commit: SystemTime,
+    offsets: HashMap<Partition, Position>,
 }
 impl ProcessingStrategy<KafkaPayload> for Next {
     fn poll(&mut self) -> Option<CommitRequest> {
+        let now = SystemTime::now();
+        let diff = now.duration_since(self.last_commit).unwrap();
+        if diff > COMMIT_INTERVAL && self.offsets.keys().len() > 0 {
+            let positions = mem::take(&mut self.offsets);
+
+            return Some(CommitRequest { positions });
+        }
+
         None
     }
 
-    fn submit(&mut self, message: Message<KafkaPayload>) -> Result<(), MessageRejected> {
-        self.producer.produce(&self.destination, &message.payload);
+    fn submit(&mut self, message: Message<KafkaPayload>) -> Result<(), ProcessingError> {
+        self.offsets.insert(
+            message.partition,
+            Position {
+                offset: message.offset,
+                timestamp: message.timestamp,
+            },
+        );
+
+        let res = self.producer.produce(&self.destination, &message.payload);
+
+        // TODO: MessageRejected should be handled by the StreamProcessor but
+        // is not currently.
+        if let Err(ProducerError::QueueFull) = res {
+            return Err(ProcessingError::MessageRejected);
+        }
```

Comment:
Do you plan to start using futures instead? The issue with this approach, I think, is that it puts you in a CPU tight loop while you wait for the message to be produced, while you can leave the CPU free by waiting on a fuiture.

Reply:
I think using futures might make more sense. But this is not actually a synchronous function currently and we are not waiting. It fires a delivery callback when it's done.